src/query/api/v1/handler/influxdb/write.go

```diff
@@ -188,11 +188,26 @@ func (ii *ingestIterator) Next() bool {
 	return false
 }
 
+func copyTagsWithNewName(t models.Tags, name []byte) models.Tags {
+	copiedTags := make([]models.Tag, t.Len())
+	metricName := t.Opts.MetricName()
+	nameHandled := false
+	for i, tag := range t.Tags {
+		if !nameHandled && bytes.Equal(tag.Name, metricName) {
+			copiedTags[i] = models.Tag{Name: metricName, Value: name}
+			nameHandled = true
+		} else {
+			copiedTags[i] = tag
+		}
+	}
+	return models.Tags{Tags: copiedTags, Opts: t.Opts}
+}
+
 func (ii *ingestIterator) Current() (models.Tags, ts.Datapoints, xtime.Unit, []byte) {
 	if ii.pointIndex < len(ii.points) && ii.nextFieldIndex > 0 && len(ii.fields) > (ii.nextFieldIndex-1) {
 		point := ii.points[ii.pointIndex]
 		field := ii.fields[ii.nextFieldIndex-1]
-		tags := ii.tags.SetName(field.name)
+		tags := copyTagsWithNewName(ii.tags, field.name)
 
 		return tags, []ts.Datapoint{ts.Datapoint{Timestamp: point.Time(),
 			Value: field.value}}, xtime.Nanosecond, nil
```

Comment:
Note: I missed this before, it might be worth trying to choose the most granular precision to improve the M3TSZ compression ratio.

Otherwise there could be high variability in the delta-of-delta encoding when not required (i.e. a delta-of-delta encoding 4000000 nanoseconds takes many more bits than a delta-to-delta encoding of 4 milliseconds).

You could basically divide each time unit by the nanoseconds and see which one falls within exactly each unit then choose the highest occuring one that equally divides (i.e. try seconds first, then millisecond, then so on).  If they exactly fall into a unit, then emit with that unit?

Reply:
I actually wondered about that, because we get pretty lousy M3TSZ compression ratio. That would explain it, I will address that too.

Reply:
Added it in commit, but it's only ~10% space savings in our test load (1/N scale of real thing). However, better than nothing.